Resource hints 2021 markdown (#2452)
HTTPArchive/almanac.httparchive.org: src/content/en/2021/resource-hints.md

---
#See https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Authors'-Guide#metadata-to-add-at-the-top-of-your-chapters
title: Resource Hints
description: Resource Hints cahpter of the 2021 Web Almanac covering adoption of resource hints, their uses, bad practices and their impact on performance.
authors: [kevinfarrugia]
reviewers: [siakaramalegos, tunetheweb, samarpanda]
analysts: [nithanaroy]
editors: []
translators: []
results: https://docs.google.com/spreadsheets/d/1Mw6TjkIClRtlZPHbij5corOZbaSUp-vgTVq3Ig18IwQ/
kevinfarrugia_bio: Kevin Farrugia is a consultant on web performance and software architecture. You can find him blogging on <a hreflang="en" href="https://imkev.dev">imkev.dev</a>.
featured_quote: Resource hints could become a double-edged sword if the benefit for your users is not evaluated. A quarter of preload requests went unused while the number of preload hits correlated with slower load times.
featured_stat_1: 44.28%
featured_stat_label_1: Adoption of preload in the top 1,000 websites.
featured_stat_2: 21.48%
featured_stat_label_2: Unused preload hints within the first 3 seconds.
featured_stat_3: 60%

> featured_stat_label_3: Lazy loaded images within the initial viewport.

Author: Updated.

---

## Introduction

> Resource Hints are instructions to the browser that you may use to improve a website's performance. This set of instructions enables you to assist the browser in prioritizing origins or resources which need to be fetched and processed.

Author: Want to keep the intro as simple as possible tbh. The use of resource hints is still to improve a website's performance.

**HTTP Header**

```
Link: <https://example.com\>; rel=dns-prefetch
```

**JavaScript**

```
const link = document.createElement("link");
link.rel="prefetch";
link.href="https://example.com";

document.head.appendChild(link);
```

{{ figure_markup(
  image="http-headers-vs-html-markup.png",
  caption="Popularity of resource hints as HTTP Headers and HTML markup.",
  description="Bar chart showing the popularity of resource hints as HTTP Headers and HTML markup. HTML markup is used on 91.24% of instances on desktop and 94.13% of instances on mobile.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTgxCYoH2vDbhTky1qQKEeV034kReHF8JYYq0aDyoo6LG22XL6Tar3dNPsqU1_zyvR6vuFEdMGoF1oP/pubchart?oid=1056457221&format=interactive",
  sheets_gid="1707816066",
  sql_file="http_headers_hints_adoption.sql"
  )
}}

Adoption for HTTP Headers is significantly lower than having resource hints implemented as part of the document markup; with less than 1.5% of the pages analyzed implementing resource hints through HTTP Headers. This is likely attributed to the ease with which they may be added or modified from within the HTML source, when compared to adding an HTTP header on the server.

> Using our current methodology, it is not possible to reliably measure resource hints which are added following user-interaction, such as <a hreflang="en" href="https://github.com/GoogleChromeLabs/quicklink">QuickLink</a>, which featured on less than 0.1% of pages analyzed.

Reviewer: Tweaked wording slightly.

{{ figure_markup(
  image="rel-dns-prefetch-adoption-by-rank.png",
  caption="Adoption of rel=\"dns-prefetch\" segmented by CrUX rank.",
  description="Bar chart showing the adoption of rel=\"dns-prefetch\" segmented by CrUX rank. rel=\"dns-prefetch\" has a 29% for the top 1,000 sites on mobile and 28% on desktop. rel=\"dns-prefetch\" has a 25% for the top 10,000 sites on mobile and 25% on desktop. rel=\"dns-prefetch\" has a 23% for the top 100,000 sites on mobile and 23% on desktop. rel=\"dns-prefetch\" has a 27% for the top 1 million sites on mobile and 27% on desktop. rel=\"dns-prefetch\" has a 36% for the top 10 million sites on mobile and 36% on desktop.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTgxCYoH2vDbhTky1qQKEeV034kReHF8JYYq0aDyoo6LG22XL6Tar3dNPsqU1_zyvR6vuFEdMGoF1oP/pubchart?oid=1340595902&format=interactive",
  sheets_gid="1880502987",
  sql_file="hints_adoption_by_rank.sql"
  )
}}

`dns-prefetch` is the only resource hint which exhibits a decrease in adoption when comparing the top 1,000 sites with the overall adoption.

{{ figure_markup(
  image="rel-preconnect-adoption-by-rank.png",

>   caption="Adoption of rel=\"preconnect\" segmented by CrUX rank.",
>   description="Bar chart showing the adoption of rel=\"preconnect\" segmented by CrUX rank. rel=\"preconnect\" has a 30% for the top 1,000 sites on mobile and 29% on desktop. rel=\"preconnect\" has a 25% for the top 10,000 sites on mobile and 25% on desktop. rel=\"preconnect\" has a 18% for the top 100,000 sites on mobile and 18% on desktop. rel=\"preconnect\" has a 14% for the top 1 million sites on mobile and 14% on desktop. rel=\"preconnect\" has a 13% for the top 10 million sites on mobile and 13% on desktop.",

Reviewer: Code format. Can also switch to single quotes to avoid having to escape:

```suggestion
  caption='Adoption of `rel="preconnect"` segmented by CrUX rank.',
  description='Bar chart showing the adoption of `rel="preconnect"` segmented by CrUX rank. `rel="preconnect"` has a 30% for the top 1,000 sites on mobile and 29% on desktop. `rel="preconnect"` has a 25% for the top 10,000 sites on mobile and 25% on desktop. `rel="preconnect"` has a 18% for the top 100,000 sites on mobile and 18% on desktop. `rel="preconnect"` has a 14% for the top 1 million sites on mobile and 14% on desktop. `rel="preconnect"` has a 13% for the top 10 million sites on mobile and 13% on desktop.',
```

  sheets_gid="1880502987",
  sql_file="hints_adoption_by_rank.sql"
  )
}}

To counter this decrease, the top 1,000 pages have an increased adoption for the `preconnect` hint taking advantage of the benefits covered earlier on in the chapter.

### Usage

Resource hints can be very effective if used correctly. By shifting the responsibility from the browser to the developer, it allows you to prioritise resources required for the critical rendering path and improve the load times & user experience.

<figure markdown>
| rank       | preload | prefetch | preconnect | prerender | dns-prefetch | modulepreload |
|------------|---------|----------|------------|-----------|--------------|-------------- |
| 1,000      |       3 |        2 |          4 |         0 |            4 |             1 |
| 10,000     |       3 |        1 |          4 |         1 |            3 |             1 |
| 100,000    |       2 |        2 |          3 |         1 |            3 |             1 |
| 1,000,000  |       2 |        2 |          2 |         1 |            2 |             1 |

> | 10,000,000 |       2 |        2 |          1 |         1 |            2 |             1 |

Reviewer: Since we have less than 10million sites we should avoid labelling that bucket as such.

```suggestion
| all |       2 |        2 |          1 |         1 |            2 |             1 |
```

{{ figure_markup(
  image="preload-as-attribute-values.png",
  caption="rel=\"preload\" as attribute values.",
  description="Bar chart showing the usage of values for rel=\"preload\". script is used 55% of the time on mobile and 54% on desktop. font is used 22% of the time on mobile and 23% on desktop. style is used 11% of the time on mobile and 11% on desktop. fetch is used 9% of the time on mobile and 8% on desktop. image is used 3% of the time on mobile and 3% on desktop.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTgxCYoH2vDbhTky1qQKEeV034kReHF8JYYq0aDyoo6LG22XL6Tar3dNPsqU1_zyvR6vuFEdMGoF1oP/pubchart?oid=1844544440&format=interactive",
  sheets_gid="1246058294",
  sql_file="hint_attribute_usage.sql"
  )
}}

> ##### script

Reviewer: ```suggestion
##### `script`
```

`script` is the most common value by a significant margin. `<script>` elements are usually discovered early as they are embedded in the initial HTML document, but it is a common practice to place `<script>` elements before the closing `<body>` tag. Since HTML is parsed sequentially, this means that the scripts will be discovered after the DOM is downloaded and parsed - and with more websites dependent on JavaScript frameworks, the necessity to have JavaScript load early has increased. The downside is that JavaScript resources would be prioritized over the other resources discovered within the HTML document, including images and stylesheets, possibly compromising the user experience.

> ##### font

Reviewer: ```suggestion
##### `font`
```

The second most commonly preloaded resource is the `font`, which is a late-discovered resource since the browser will only download a font file after the layout phase when the browser knows that the font will be rendered on the page.

> ##### style

Reviewer: ```suggestion
##### `style`
```

Stylesheets are ordinarily embedded in the document's `<head>` and discovered early during the document parsing. Additionally, as stylesheets are render-blocking resources they are assigned the _Highest_ request priority. This should make preloading stylesheets unnecessary but it is sometimes required to re-prioritize the requests. A <a hreflang="en" href="https://bugs.chromium.org/p/chromium/issues/detail?id=629420">bug</a> in Google Chrome (fixed in Chrome 95) prioritizes preloaded resources ahead of other higher-priority resources discovered by the preload scanner, including CSS files. Preloading the stylesheet will restore its _Highest_ priority. Another instance when stylesheets are preloaded is when they are not downloaded directly from the HTML document, such as the <a hreflang="en" href="https://www.filamentgroup.com/lab/async-css.html">asynchronous CSS</a> "hack" which uses an `onload` event to avoid render-blocking the page with non-critical CSS.

> ##### fetch

Reviewer: ```suggestion
##### `fetch`
```

Preload may also be used to initiate a request to retrieve data which you know is critical to the rendering of the page, such as a JSON response or stream.

> ##### image

Reviewer: ```suggestion
##### `image`
```

Preloading images may be helpful in improving the LCP score when the image is not included in the initial HTML, such as a CSS `background-image`.

#### The `crossorigin` attribute

The `crossorigin` attribute is used to indicate whether CORS must be used when fetching the requested resource. This could apply to any resource type, but it is most commonly associated with font files as they should always be requested using CORS.

> <figure markdown>
> | value                       | desktop | mobile | 
> |-----------------------------|---------|--------| 
> | not set                     | 66.61%  | 65.91% | 
> | crossorigin (or equivalent) | 14.50%  | 13.48% | 
> | use-credentials             | 0.04%   | 0.04%  | 
> 

Reviewer: We convert tables with percentages to HTML to allow us to apply a `class="numeric"` attribute to those `<td>`s to right-align them:

```suggestion
<figure markdown>
<table>
  <thead>
    <tr>
      <th>value</th>
      <th>desktop</th>
      <th>mobile</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>not set</td>
      <td class="numeric">66.61%</td>
      <td class="numeric">65.91%</td>
    </tr>
    <tr>
      <td><code>crossorigin</code> (or equivalent)</td>
      <td class="numeric">14.50%</td>
      <td class="numeric">13.48%</td>
    </tr>
    <tr>
      <td><code>use-credentials</code></td>
      <td class="numeric">0.04%</td>
      <td class="numeric">0.04%</td>
    </tr>
  </tbody>
</table>
```

> <figcaption>{{ figure_link(caption="rel=\"preload\" crossorigin attribute values.", sheets_gid="1246058294", sql_file="hint_attribute_usage.sql") }}</figcaption>

Reviewer: ```suggestion
<figcaption>{{ figure_link(caption='`rel="preload"` `crossorigin` attribute values.', sheets_gid="1246058294", sql_file="hint_attribute_usage.sql") }}</figcaption>
```

</figure>

##### anonymous

The default value when no value is specified, `anonymous` will set the credentials flag to <a hreflang="en" href="https://developer.mozilla.org/en-US/docs/Web/Security/Same-origin_policy">`same-origin`</a> and is required when downloading resources protected by CORS. It is also a <a hreflang="en" href="https://drafts.csswg.org/css-fonts/#font-fetching-requirements">requirement</a> when downloading font files - even if they are on the same origin! If you omit the `crossorigin` attribute when the eventual request for the preloaded resource uses CORS, you will end up with a duplicate request since it won't match in the preload cache.

> ##### use-credentials

Reviewer: ```suggestion
##### `use-credentials`
```

In addition to the `media` attribute, the `<link>` element supports `imagesrcset` and `imagesizes` attributes which correspond to the `srcset` and `sizes` attributes on `<img>` elements. Using these attributes, you are able to use the same resource selection criteria that you would use on your image. Unfortunately their adoption is very low (less than 1%); most likely owing to the lack of <a hreflang="en" href="https://caniuse.com/mdn-html_elements_link_imagesizes">support</a> on Safari.

<p class="note">Note: The <code>media</code> attribute is not available on all <code>&lt;link&gt;</code> elements as the spec suggests, but it is only available on <code>rel="preload"</code>.</p>

### Bad practices

Owing to the versatility of `rel="preload"`, there isn't a clear set of rules dictating how to implement the preload hint, but we are able to learn a lot from our mistakes and understand how to avoid them.

#### Unused preload

> We have already seen that there is a negative correlation between a website's performance and the number of preload hints. This relationship may be influenced by two factors:

Reviewer: ```suggestion
We have already seen that there is a negative correlation between a website's performance and the number of preload hints. This relationship may be influenced by two factors:
- Incorrect preloads
- Unused preloads
```

##### Incorrect preloads

When you preload a resource which is not as important as the other resources which the browser would have otherwise prioritized. We are unable to measure the extent of incorrect preloads as you would need to A/B test the page with and without each hint.

##### Unused preloads

When you preload a resource which is not needed within the first few seconds of the page loading.

{{ figure_markup(
  caption="Percent of unused preload hints within the first 3 seconds.",
  content="21.48%",
  classes="big-number",
  sheets_gid="2013605735",
  sql_file="consoleLog_unused_preload.sql"
)
}}

In such cases, the preload hint is regressing the website's performance, as you are instructing the browser to download and prioritize files or resources which are not needed immediately - or even not needed at all. This is one of the challenges when using resource hints, as they require regular maintenance and automating the process opens the door to allow such issues to creep in.

{{ figure_markup(

>   caption="The most unused rel=\"preload\" hints on a single page",

Reviewer: ```suggestion
  caption='The most unused `rel="preload"` hints on a single page',
```

There are 17,861 cases of unrecognised values, with the most frequent error being omitting it completely; while the most common invalid as values are `other` and `stylesheet` (the correct value is `style`).

{{ figure_markup(
  caption="Pages incorrectly used as=\"stylesheet\" instead of \"style\"",
  content="1,114",
  classes="big-number",
  sheets_gid="1681733418",
  sql_file="consoleLog_unused_preload.sql"
)
}}

When using an incorrect `as` attribute value - as opposed to unrecognised value, such as using `style` instead of `script`; the browser will duplicate the file download as the request won't match the resource stored in the preload cache.

<p class="note">Note: While <code>video</code> is included in the spec, it isn't supported by any browser and would be treated as an invalid value and ignored.</p>

#### Unused font files

More than 5% of pages which preload font files, preload more font files than needed. When preloading font files, all browsers which support `preload` also support `.woff2`. This means that, assuming that the `.woff2` font files are available, it is not necessary to preload older formats, including `.woff`.

### Third Parties

> You are able to use resource hints to connect to or download resources from both first and [third parties](./third-parties). While `dns-prefetch` and `preconnect` are only useful when connecting to different origins, including subdomains; `preload` and `prefetch` may be used for both resources on the same origin and resources hosted by third parties.

Reviewer: ```suggestion
You are able to use resource hints to connect to, or download resources from, both first and [third parties](./third-parties). While `dns-prefetch` and `preconnect` are only useful when connecting to different origins, including subdomains; `preload` and `prefetch` may be used for both resources on the same origin and resources hosted by third parties.
```

> When considering which resource hints you should use for third party resources, you need to evaluate the priority and role of each third party on your application's loading experience and whether the costs are justified. 

Reviewer: For adjectives we hyphenate. For nouns we don't.

```suggestion
When considering which resource hints you should use for third-party resources, you need to evaluate the priority and role of each third party on your application's loading experience and whether the costs are justified. 
```

> Prioritising third party resources over your own content is potentially a warning sign, however there are cases when this is recommended. As an example, if we look at cookie notice scripts; which are required in the European Union by General Data Protection Regulation; these are usually accompanied by a `dns-prefetch` or `preconnect` script as they are highly obtrusive to the user experience and also a prerequisite for some site functions, such as serving personalised ads.

Reviewer: ```suggestion
Prioritising third-party resources over your own content is potentially a warning sign, however there are cases when this is recommended. As an example, if we look at cookie notice scripts; which are required in the European Union by General Data Protection Regulation; these are usually accompanied by a `dns-prefetch` or `preconnect` script as they are highly obtrusive to the user experience and also a prerequisite for some site functions, such as serving personalised ads.
```

Also - add a link to GDPR? Wikipedia or the like?

<figure markdown>
| host                       | dns-prefetch | preconnect | preload | Total       | 
|----------------------------|--------------|------------|---------|-------------| 
| adservice.google.com       | 0.25%        | 0.37%      | 42.72%  | 43.34%      | 
| fonts.gstatic.com          | 0.82%        | 25.45%     | 0.74%   | 27.01%      | 
| fonts.googleapis.com       | 14.09%       | 4.59%      | 3.05%   | 21.72%      | 
| s.w.org                    | 20.00%       | 0.25%      | -       | 20.25%      | 
| cdn.shopify.com            | -            | 1.82%      | 12.16%  | 13.98%      | 
| www.google-analytics.com   | 1.27%        | 4.05%      | 0.17%   | 5.49%       | 
| www.googletagmanager.com   | 1.94%        | 3.14%      | 0.25%   | 5.32%       | 
| siteassets.parastorage.com | -            | -          | 5.24%   | 5.24%       | 
| images.squarespace-cdn.com | -            | 4.51%      | -       | 4.51%       | 
| ajax.googleapis.com        | 2.30%        | 1.79%      | 0.36%   | 4.45%       | 

> 

Reviewer: Can you convert to HTML like above? I find it best to grab HTML from Dev Tools and then add `class="numeric"` attributes.

<figcaption>{{ figure_link(caption="Most popular third parties connected to using resource hints.", sheets_gid="1254656281", sql_file="preload_host_by_url.sql") }}</figcaption>
</figure>

Analysing the table above, 42.72% of all pages which include a `preload` hint are preloading resources hosted on adservice.google.com, while s.w.org (WordPress) is the most popular domain for `dns-prefetch` and `fonts.gstatic.com` for the `preconnect` directive.

> <p class="note">Note: Google Fonts now includes instructions to <code>preconnect</code> to both the fonts.gstatic.com origin and fonts.googleapis.com.</p>

Reviewer: Would be great to get a reference, or even a screenshot for this.

> ### Native Lazy Loading

Reviewer: ```suggestion
### Native lazy-loading
```

Native image and iframe lazy-loading have been standardized in 2019 and since then, their adoption - especially for images - has grown exponentially. Lazy-loading refers to the technique to defer downloading a resource - in this case an image or iframe - until it is needed or visible within the viewport.

> `loading="lazy"` on images is supported on most major browsers. On Safari, it is marked as <a hreflang="en" href="https://bugs.webkit.org/show_bug.cgi?id=200764">in progress</a> and is available behind a flag, but not yet enabled by default. Lazy-loading of iframes has similar support, although it is <a hreflang="en" href="https://bugzilla.mozilla.org/show_bug.cgi?id=1622090">not yet supported on Firefox</a>.

Reviewer: Second sentence is a bit heard to read.

```suggestion
`loading="lazy"` on images is supported on most major browsers. On Safari, it is marked as <a hreflang="en" href="https://bugs.webkit.org/show_bug.cgi?id=200764">in progress</a> and is available behind a flag, but not yet enabled by default. Lazy-loading of iframes is supported on Chrome and again behind a flag on Safari, although it is <a hreflang="en" href="https://bugzilla.mozilla.org/show_bug.cgi?id=1622090">not yet supported on Firefox</a>.
```

> Browsers which do not support the loading attribute, will simply ignore it - making it safe to add without unwanted side-effects. JavaScript based alternatives, such as <a hreflang="en" href="https://github.com/aFarkas/lazysizes">lazysizes</a> may still be used, however considering that full browser support is around the corner, it may not be worth adding to a project at this stage.

Reviewer: ```suggestion
Browsers which do not support the `loading` attribute, will simply ignore it - making it safe to add without unwanted side-effects. JavaScript based alternatives, such as <a hreflang="en" href="https://github.com/aFarkas/lazysizes">lazysizes</a> may still be used, however considering that full browser support is around the corner, it may not be worth adding to a project at this stage.
```

{{ figure_markup(
  image="adoption-of-loading-lazy-on-img.png",

>   caption="The percent of pages that have the loading=lazy attribute on img elements.",

Reviewer: ```suggestion
  caption="The percent of pages that have the `loading=lazy` attribute on `img` elements.",
```

>   description="Time series showing the percent of pages that have the loading=lazy attribute on img elements on desktop and mobile. At 01/01/2019 the adoption rate is 0% on desktop and 0% on mobile. At 01/01/2020 the adoption rate is 0% on desktop and 0% on mobile. At 01/01/2021 the adoption rate is 14.2% on desktop and 13.5% on mobile. At 10/01/2021, the adoption rate is 19.4% on desktop and 19.4% on mobile.",

Reviewer: ```suggestion
  description="Time series showing the percent of pages that have the `loading=lazy` attribute on `img` elements on desktop and mobile. At 01/01/2019 the adoption rate is 0% on desktop and 0% on mobile. At 01/01/2020 the adoption rate is 0% on desktop and 0% on mobile. At 01/01/2021 the adoption rate is 14.2% on desktop and 13.5% on mobile. At 10/01/2021, the adoption rate is 19.4% on desktop and 19.4% on mobile.",
```

A <a hreflang="en" href="https://web.dev/lcp-lazy-loading/">study</a> on the load times for pages which use lazy-loading, indicated that pages which use lazy-loading tend to have a worse LCP performance, possibly caused by overusing the lazy-loading attribute. This is increasingly significant on the LCP element, which shouldn't be lazy-loaded. If you are using `loading="lazy"`, you should check that the lazily-loaded images are below the fold and more critically, that the [LCP element is not lazy-loaded](./performance). More than 60% of all images which are lazy-load are actually within the initial viewport and shouldn't be lazy-loaded.)

{{ figure_markup(
  image="lazy-loaded-images.png",

>   caption="Percent of img elements with loading=lazy which are in the initial viewport.",
>   description="Bar chart showing the percent of img elements with loading=lazy which are within the initial viewport. 61.54% on mobile and 63.10% on desktop of lazy-loaded images are within the initial viewport.",

Reviewer: ```suggestion
  caption="Percent of img elements with `loading=lazy` which are in the initial viewport.",
  description="Bar chart showing the percent of `img` elements with `loading=lazy` which are within the initial viewport. 61.54% on mobile and 63.10% on desktop of lazy-loaded images are within the initial viewport.",
```

  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTgxCYoH2vDbhTky1qQKEeV034kReHF8JYYq0aDyoo6LG22XL6Tar3dNPsqU1_zyvR6vuFEdMGoF1oP/pubchart?oid=977858704&format=interactive",
  sheets_gid="1468369891",
  sql_file="lazy_viewport_images.sql"
  )
}}

The likelihood of a page containing at least one iframe is much lower than for that containing an image with only 2.56% of pages containing an iframe taking advantage of native lazy-loading. The benefits of lazy-loading an iframe are potentially important, as an iframe could initiate further requests to download even more resources, including scripts and images. This is especially true when using embeds, such as YouTube or Twitter embeds. Similarly to when deciding the loading strategy for an image, you must check whether the iframe is shown within the initial viewport or not. If it isn't, then it is usually safe to add `loading="lazy"` to the `<iframe>` element to benefit from a reduced initial load & boost performance.

### HTTP/2 Server Push

> HTTP/2 supports a technology called Server Push that preemptively pushes a resource it expects the client will be requesting. As the server is pushing the resource instead of informing the client that it should request it, cache-management becomes complex and in some cases the pushed resources would even delay the delivery of the HTML, which is critical for discovering all resources required to load the page.

Reviewer: ```suggestion
HTTP/2 supports a technology called _Server Push_ that preemptively pushes a resource it expects the client will be requesting. As the server is pushing the resource instead of informing the client that it should request it, cache-management becomes complex and in some cases the pushed resources would even delay the delivery of the HTML, which is critical for discovering all resources required to load the page.
```

> HTTP/2 supports a technology called Server Push that preemptively pushes a resource it expects the client will be requesting. As the server is pushing the resource instead of informing the client that it should request it, cache-management becomes complex and in some cases the pushed resources would even delay the delivery of the HTML, which is critical for discovering all resources required to load the page.

Reviewer: ```suggestion
HTTP/2 supports a technology called _Server Push_ that preemptively pushes a resource it expects the client will be requesting. As the server is pushing the resource instead of informing the client that it should request it, cache-management becomes complex and in some cases the pushed resources would even delay the delivery of the HTML, which is critical for discovering all resources required to load the page.
```

Unfortunately HTTP/2 push has been disappointing with little evidence that it provides the performance boost promised, compared to the risk of over pushing resources that either the browser already has, or that are of less importance than resources the browser requests.

> So while the technology is widely available, overcoming these obstacles makes it highly unpopular - with less than 1% adoption on Chrome. Chrome has also filed an <a hreflang="en" href="https://lists.w3.org/Archives/Public/ietf-http-wg/2019JulSep/0078.html">Intent to Remove</a> that is paused until a testable implementation of 103 Early Hints is available.

Reviewer: Worth mentioning that Chrome also does not support Push for their HTTP/3 implementation (and neither do others like Cloudflare).

### Future

While there are no proposals to add new rel directives, improvements from the browser vendors to the current set of resource hints - such as the prioritization <a hreflang="en" href="https://bugs.chromium.org/p/chromium/issues/detail?id=629420">bug</a> in Chrome - are expected to have a positive impact. Hint adoption is expected to evolve and the use of `preload` should shift towards its intended purpose - late discovered resources.

> Additionally, two new proposals, 103 Early Hints and Priority Hints are expected to be made available in the near future, with experimental support already available on Chrome.

Reviewer: Not sure I'd call either "new". Been around for a while but only recently getting adopted. Consider rephrasing.

featured_stat_label_2: Unused preload hints within the first 3 seconds.
featured_stat_3: 60%
featured_stat_label_3: Lazy loaded images are within the initial viewport.
---

## Introduction

Resource Hints are instructions to the browser that you may use to improve a website's performance. This set of instructions enable you to assist the browser in prioritizing origins or resources which need to be fetched and processed.

Let's take a closer look at how resource hints are implemented, what are the most common pitfalls and what we can do to make sure we are using resource hints as effectively as possible. 

### The Link directive

The most widely adopted resource hints are implemented through the Link directive's rel attribute. These are <a hreflang="en" href="https://www.w3.org/TR/resource-hints/#dfn-dns-prefetch">`dns-prefetch`</a>, <a hreflang="en" href="https://www.w3.org/TR/resource-hints/#dfn-preconnect">`preconnect`</a>, <a hreflang="en" href="https://www.w3.org/TR/resource-hints/#dfn-prefetch">`prefetch`</a>, <a hreflang="en" href="https://www.w3.org/TR/resource-hints/#dfn-prerender">`prerender`</a> and <a hreflang="en" href="https://www.w3.org/TR/preload/">`preload`</a>.

> These may implemented in one of three ways:

Reviewer: ```suggestion
These may be implemented in one of three ways:
```

  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTgxCYoH2vDbhTky1qQKEeV034kReHF8JYYq0aDyoo6LG22XL6Tar3dNPsqU1_zyvR6vuFEdMGoF1oP/pubchart?oid=1056457221&format=interactive",
  sheets_gid="1707816066",
  sql_file="http_headers_hints_adoption.sql"
  )
}}

Using our current [methodology](./methodology), it is not possible to reliably measure resource hints that are added following user-interaction, such as those added through <a hreflang="en" href="https://github.com/GoogleChromeLabs/quicklink">QuickLink</a>, which featured on less than 0.1% of pages analyzed.

Considering that the adoption of resource hints using HTTP headers is markedly smaller than adoption for the `<link>` HTML element, the rest of this chapter will focus on analyzing the usage of resource hints through the HTML Element.

#### Types of Resource Hints

There are five resource hint link relationships supported by most browsers today: `dns-prefetch`, `preconnect`, `prefetch`, `prerender` and `preload`.

##### dns-prefetch

> The `dns-prefetch` link initiates an early request to resolve a domain name. It is only effective for DNS lookups on cross-origin domains and may be paired together with `preconnect`. While Chrome now supports a maximum of <a hreflang="en" href="https://source.chromium.org/chromium/chromium/src/+/fdf9418d23d434e0f7134da67dc41b0fe8268e91:net/dns/host_resolver_manager.cc;l=416">64</a> concurrent in-flight DNS requests - up from 6 last year - other browsers still have tighter limitations. For example it is limited to <a hreflang="en" href="https://github.com/mozilla/gecko-dev/blob/master/netwerk/dns/nsHostResolver.h#L48">8</a> on Firefox.

Reviewer: ```suggestion
The `dns-prefetch` link initiates an early request to resolve a domain name. It is only effective for DNS lookups on cross-origin domains and may be paired together with `preconnect`. While Chrome now supports a maximum of <a hreflang="en" href="https://source.chromium.org/chromium/chromium/src/+/fdf9418d23d434e0f7134da67dc41b0fe8268e91:net/dns/host_resolver_manager.cc;l=416">64</a> concurrent in-flight DNS requests - up from 6 last year - other browsers still have tighter limitations. For example, it is limited to <a hreflang="en" href="https://github.com/mozilla/gecko-dev/blob/master/netwerk/dns/nsHostResolver.h#L48">8</a> on Firefox.
```

```html
<link rel="dns-prefetch" href="https://example.com/">
```

##### preconnect

> The `preconnect` link behaves similarly to `dns-prefetch`, but in addition to DNS lookups, it also establishes a connection together with TLS handshake if served over HTTPS. You are able to use `preconnect` in place of `dns-prefetch` as it gives a greater performance boost; but you must use it sparingly as each certificate weighs at least 3KB, which would be competing with bandwidth for other resources. You also want to avoid wasting CPU time opening connections which aren't required for critical resource. Keep in mind that if a connection isn't used within a short period of time (e.g. 10 seconds on Chrome), it would automatically be closed by the browser.

Author: Word on the street. :) 

Will update.

### Usage

Resource hints can be very effective if used correctly. By shifting the responsibility from the browser to the developer, it allows you to prioritise resources required for the critical rendering path and improve the load times & user experience.

<figure markdown>
| rank       | preload | prefetch | preconnect | prerender | dns-prefetch | modulepreload |
|------------|---------|----------|------------|-----------|--------------|-------------- |
| 1,000      |       3 |        2 |          4 |         0 |            4 |             1 |
| 10,000     |       3 |        1 |          4 |         1 |            3 |             1 |
| 100,000    |       2 |        2 |          3 |         1 |            3 |             1 |
| 1,000,000  |       2 |        2 |          2 |         1 |            2 |             1 |
| all        |       2 |        2 |          1 |         1 |            2 |             1 |

<figcaption>{{ figure_link(caption='Resource hint adoption by rank.', sheets_gid="528380369", sql_file="resource_hints_distribution_by_rank.sql") }}</figcaption>
</figure>

> Of the sites using resource hints, when comparing the median for the top 1,000 sites to the entire corpus, the top ranking sites have more resource hints per page. The only hint which observes a different pattern is `prerender`; which has a total of 0 occurrences in the top 1,000 sites.

Reviewer: ```suggestion
Of the sites using resource hints, when comparing the median for the top 1,000 sites to the entire corpus, the top-ranking sites have more resource hints per page. The only hint which observes a different pattern is `prerender`; which has a total of 0 occurrences in the top 1,000 sites.
```

While this doesn't prove that having preload hints causes a page to get slower, having many does coincide with having slower performance. Every page has its unique requirements and it is impossible to apply a "one size fits all" approach, but in the majority of cases the number of preloaded resources should be kept low and resource prioritization should be delegated to the browser when possible.

<p class="note">Note: In addition to the number of hints, the size of each preloaded resource has an impact on the website performance. The above figure does not take into consideration the size of each preloaded resource.</p>

### `rel="preload"`

With that being said, and the expectation that more websites will adopt `preload`, let's take a better look at the preload resource hint and understand why it is so effective, yet at the same time so prone to misuse.

#### The `as` attribute

> The `as` attribute should be specified when using `rel="preload"` (or `rel="prefetch"`) to specify the type of resource being downloaded. Applying the correct `as` attribute allows the browser to prioritize the resource more accurately. For example, preload as="script" will get a low or medium priority, while preload as="style" would get the highest priority. The `as` attribute is required for caching the resource for future requests and applying the correct <a hreflang="en" href="https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP">content-security policy</a>.

Reviewer: ```suggestion
The `as` attribute should be specified when using `rel="preload"` (or `rel="prefetch"`) to specify the type of resource being downloaded. Applying the correct `as` attribute allows the browser to prioritize the resource more accurately. For example, `preload as="script"` will get a low or medium priority, while `preload as="style"` would get the highest priority. The `as` attribute is required for caching the resource for future requests and applying the correct <a hreflang="en" href="https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP">Content Security Policy</a>.
```

{{ figure_markup(
  image="preload-as-attribute-values.png",

>   caption='rel="preload" as attribute values.',

Reviewer: ```suggestion
  caption='`el="preload" as` attribute values.',
```

  sheets_gid="1246058294",
  sql_file="hint_attribute_usage.sql"
  )
}}

##### `script`

`script` is the most common value by a significant margin. `<script>` elements are usually discovered early as they are embedded in the initial HTML document, but it is a common practice to place `<script>` elements before the closing `<body>` tag. Since HTML is parsed sequentially, this means that the scripts will be discovered after the DOM is downloaded and parsed - and with more websites dependent on JavaScript frameworks, the necessity to have JavaScript load early has increased. The downside is that JavaScript resources would be prioritized over the other resources discovered within the HTML document, including images and stylesheets, possibly compromising the user experience.

##### `font`

The second most commonly preloaded resource is the `font`, which is a late-discovered resource since the browser will only download a font file after the layout phase when the browser knows that the font will be rendered on the page.

##### `style`

> Stylesheets are ordinarily embedded in the document's `<head>` and discovered early during the document parsing. Additionally, as stylesheets are render-blocking resources they are assigned the _Highest_ request priority. This should make preloading stylesheets unnecessary but it is sometimes required to re-prioritize the requests. A <a hreflang="en" href="https://bugs.chromium.org/p/chromium/issues/detail?id=629420">bug</a> in Google Chrome (fixed in Chrome 95) prioritizes preloaded resources ahead of other higher-priority resources discovered by the preload scanner, including CSS files. Preloading the stylesheet will restore its _Highest_ priority. Another instance when stylesheets are preloaded is when they are not downloaded directly from the HTML document, such as the <a hreflang="en" href="https://www.filamentgroup.com/lab/async-css.html">asynchronous CSS</a> "hack" which uses an `onload` event to avoid render-blocking the page with non-critical CSS.

Reviewer: ```suggestion
Stylesheets are ordinarily embedded in the document's `<head>` and discovered early during the document parsing. Additionally, as stylesheets are render-blocking resources they are assigned the _Highest_ request priority. This should make preloading stylesheets unnecessary, but it is sometimes required to re-prioritize the requests. A <a hreflang="en" href="https://bugs.chromium.org/p/chromium/issues/detail?id=629420">bug</a> in Google Chrome (fixed in Chrome 95) prioritizes preloaded resources ahead of other higher-priority resources discovered by the preload scanner, including CSS files. Preloading the stylesheet will restore its _Highest_ priority. Another instance when stylesheets are preloaded is when they are not downloaded directly from the HTML document, such as the <a hreflang="en" href="https://www.filamentgroup.com/lab/async-css.html">asynchronous CSS</a> "hack" which uses an `onload` event to avoid render-blocking the page with non-critical CSS.
```

Preload may be used to initiate a request to retrieve data which you know is critical to the rendering of the page, such as a JSON response or stream.

##### `image`

Preloading images may help improve the LCP score when the image is not included in the initial HTML, such as a CSS `background-image`.

#### The `crossorigin` attribute

The `crossorigin` attribute is used to indicate whether CORS must be used when fetching the requested resource. This could apply to any resource type, but it is most commonly associated with font files as they should always be requested using CORS.

<figure>
  <table>
    <thead>
      <tr>
        <th>value</th>
        <th>desktop</th>
        <th>mobile</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>not set</td>
        <td class="numeric">66.61%</td>
        <td class="numeric">65.91%</td>
      </tr>
      <tr>
        <td><code>crossorigin</code> (or equivalent)</td>
        <td class="numeric">14.50%</td>
        <td class="numeric">13.48%</td>
      </tr>
      <tr>
        <td><code>use-credentials</code></td>
        <td class="numeric">0.04%</td>
        <td class="numeric">0.04%</td>
      </tr>
    </tbody>
  </table>

> <figcaption>{{ figure_link(caption='`rel="preload"` `crossorigin` attribute values.', sheets_gid="1246058294", sql_file="hint_attribute_usage.sql") }}</figcaption>

Reviewer: This isn't generating properly. Fill come back to it and investigate why.

Author: Fixed. Need to use `<code>` since this isn't markdown.

Reviewer: Yeah it should support both though. But let's go with that for the moment.

Reviewer: Ah it's cause it's in the `<figcaption>` tag within a `<table>` — unlike a graph embed which lives outside any HTML elements. OK that makes sense. Will still see if we can fix though as not totally obvious.

Author: I think if i use `<figure markdown>` as the opening element it would work, but since this is an HTML table I removed `markdown`

Reviewer: Fixed this in https://github.com/HTTPArchive/almanac.httparchive.org/pull/2458 . Up to you whether you wanna switch back to markup or happy to leave as HTML.

##### `use-credentials`

When requesting cross-origin resources which require authentication, for example through the use of cookies, client certificates or the `Authorization` header; setting the `crossorigin="use-credentials"` attribute will include this data in the request and allowing the server to respond to the request so that the resource may be preloaded. This is not a common scenario with 0.04% usage, however if your page content is dependent on an authenticated status, it could be used to initiate an early fetch request to get the login status.

#### The `media` attribute

An oft neglected feature available to `rel="preload"`, is the ability to specify media queries through the `media` attribute - with less than 4% of all preloads using this attribute. The `media` attribute accepts media queries allowing you to target the media type and specific browser features, such as viewport width. As an example, the `media` attribute would allow you to preload a low-resolution image on devices with a narrow viewport and a full-sized image on devices with a large viewport.

> In addition to the `media` attribute, the `<link>` element supports `imagesrcset` and `imagesizes` attributes which correspond to the `srcset` and `sizes` attributes on `<img>` elements. Using these attributes, you can use the same resource selection criteria that you would use on your image. Unfortunately their adoption is very low (less than 1%); most likely owing to the lack of <a hreflang="en" href="https://caniuse.com/mdn-html_elements_link_imagesizes">support</a> on Safari.

Reviewer: ```suggestion
In addition to the `media` attribute, the `<link>` element supports `imagesrcset` and `imagesizes` attributes which correspond to the `srcset` and `sizes` attributes on `<img>` elements. Using these attributes, you can use the same resource selection criteria that you would use on your image. Unfortunately, their adoption is very low (less than 1%); most likely owing to the lack of <a hreflang="en" href="https://caniuse.com/mdn-html_elements_link_imagesizes">support</a> on Safari.
```

  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTgxCYoH2vDbhTky1qQKEeV034kReHF8JYYq0aDyoo6LG22XL6Tar3dNPsqU1_zyvR6vuFEdMGoF1oP/pubchart?oid=277634745&format=interactive",
  sheets_gid="699889350",
  sql_file="consoleLog_incorrect_crossorigin_type.sql"
  )
}}

More than half of the cases when the `crossorigin` attribute on the `rel="preload"` hint is either missing or incorrect are linked to the preloading of font files, with a total of 14,818 instances across the dataset.

#### Invalid `as` attribute

The `as` attribute plays an important role when preloading your resources and getting this wrong may result in downloading the same resource twice. On most browsers, specifying an unrecognized `as` attribute will ignore the preload. The supported values are `audio`, `document`, `embed`, `fetch`, `font`, `image`, `object`, `script`, `style`, `track`, `worker` and `video`. 

There are 17,861 cases of unrecognised values, with the most frequent error being omitting it completely; while the most common invalid as values are `other` and `stylesheet` (the correct value is `style`).

{{ figure_markup(

>   caption='Pages incorrectly used as="stylesheet" instead of "style"',

Reviewer: ```suggestion
  caption='Pages incorrectly used `as="stylesheet"` instead of `"style"`',
```

      </tr>
      <tr>
        <td>images.squarespace-cdn.com</td>
        <td class="numeric">-</td>
        <td class="numeric">4.51%</td>
        <td class="numeric">-</td>
        <td class="numeric">4.51%</td>
      </tr>
      <tr>
        <td>ajax.googleapis.com</td>
        <td class="numeric">2.30%</td>
        <td class="numeric">1.79%</td>
        <td class="numeric">0.36%</td>
        <td class="numeric">4.45%</td>
      </tr>
    </tbody>
  </table>

<figcaption>{{ figure_link(caption="Most popular third parties connected to using resource hints.", sheets_gid="1254656281", sql_file="preload_host_by_url.sql") }}</figcaption>
</figure>

Analysing the table above, 42.72% of all pages which include a `preload` hint are preloading resources hosted on adservice.google.com, while s.w.org (WordPress) is the most popular domain for `dns-prefetch` and `fonts.gstatic.com` for the `preconnect` directive.

{{ figure_markup(
  image="google-fonts.png",
  alt="Google Fonts instructions to embed a font.",
  caption='Google Fonts instructions to preconnect to fonts.gstatic.com and fonts.googleapis.com. From <a hreflang="en" href="https://fonts.google.com/">Google Fonts</a>.',

>   description='Code snippet titled "Use on the web" and preded with "To embed a font, copy the code into the &lt;head&gt; of your html". Code snippet states `&lt;link rel="preconnect" href="https://fonts.googleapis.com"&gt;&lt;link rel="preconnect" href="https://fonts.gstatic.com" crossorigin&gt;&lt;link href="https://fonts.googleapis.com/css2?family=Roboto&display=swap" rel="stylesheet"&gt;`',

Reviewer: Preded?

Author: What's preded?

Reviewer: That's my question :-)

> Code snippet titled "Use on the web" and preded with "To embed a font, copy the code into the &lt;head&gt; of your html

{{ figure_markup(
  image="lazy-loaded-images.png",
  caption='Percent of `img` elements with `loading="lazy"` which are in the initial viewport.',
  description='Bar chart showing the percent of `img` elements with `loading="lazy"` which are within the initial viewport. 61.54% on mobile and 63.10% on desktop of lazy-loaded images are within the initial viewport. 38.46% on mobile and 36.90% on desktop of lazy-loaded images are not within the initial viewport.',
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTgxCYoH2vDbhTky1qQKEeV034kReHF8JYYq0aDyoo6LG22XL6Tar3dNPsqU1_zyvR6vuFEdMGoF1oP/pubchart?oid=977858704&format=interactive",
  sheets_gid="1468369891",
  sql_file="lazy_viewport_images.sql"
  )
}}

> The likelihood of a page containing at least one iframe is much lower than for that containing an image with only 2.56% of pages containing an iframe taking advantage of native lazy-loading. The benefits of lazy-loading an iframe are potentially important, as an iframe could initiate further requests to download even more resources, including scripts and images. This is especially true when using embeds, such as YouTube or Twitter embeds. Similarly to when deciding the loading strategy for an image, you must check whether the iframe is shown within the initial viewport or not. If it isn't, then it is usually safe to add `loading="lazy"` to the `<iframe>` element to benefit from a reduced initial load & boost performance.

Reviewer: ```suggestion
The likelihood of a page containing at least one iframe is much lower than for that containing an image with only 2.56% of pages containing an iframe taking advantage of native lazy-loading. The benefits of lazy-loading an iframe are potentially important, as an iframe could initiate further requests to download even more resources, including scripts and images. This is especially true when using embeds, such as YouTube or Twitter embeds. Similarly, to when deciding the loading strategy for an image, you must check whether the iframe is shown within the initial viewport or not. If it isn't, then it is usually safe to add `loading="lazy"` to the `<iframe>` element to benefit from a reduced initial load & boost performance.
```

### HTTP/2 Server Push

> HTTP/2 supports a technology called _Server Push_ that preemptively pushes a resource it expects the client will be requesting. As the server is pushing the resource instead of informing the client that it should request it, cache-management becomes complex and in some cases the pushed resources would even delay the delivery of the HTML, which is critical for discovering all resources required to load the page.

Reviewer: ```suggestion
HTTP/2 supports a technology called _Server Push_ that preemptively pushes a resource it expects the client will be requesting. As the server is pushing the resource instead of informing the client that it should request it, cache-management becomes complex and, in some cases, the pushed resources would even delay the delivery of the HTML, which is critical for discovering all resources required to load the page.
```

> Unfortunately HTTP/2 push has been disappointing with little evidence that it provides the performance boost promised, compared to the risk of over pushing resources that either the browser already has, or that are of less importance than resources the browser requests.

Reviewer: ```suggestion
Unfortunately, HTTP/2 push has been disappointing with little evidence that it provides the performance boost promised, compared to the risk of over pushing resources that either the browser already has, or that are of less importance than resources the browser requests.
```

> So while the technology is widely available, overcoming these obstacles makes it highly unpopular - with less than 1% adoption on Chrome. Chrome has also filed an <a hreflang="en" href="https://lists.w3.org/Archives/Public/ietf-http-wg/2019JulSep/0078.html">Intent to Remove</a> that is paused until a testable implementation of 103 Early Hints is available. Chrome <a hreflang="en" href="https://github.com/httpwg/http2-spec/issues/786#issuecomment-724371629">does not support</a> _Server Push_ on HTTP/3 either.

Reviewer: ```suggestion
So, while the technology is widely available, overcoming these obstacles makes it highly unpopular - with less than 1% adoption on Chrome. Chrome has also filed an <a hreflang="en" href="https://lists.w3.org/Archives/Public/ietf-http-wg/2019JulSep/0078.html">Intent to Remove</a> that is paused until a testable implementation of 103 Early Hints is available. Chrome <a hreflang="en" href="https://github.com/httpwg/http2-spec/issues/786#issuecomment-724371629">does not support</a> _Server Push_ on HTTP/3 either.
```

### Future

> While there are no proposals to add new rel directives, improvements from the browser vendors to the current set of resource hints - such as the prioritization <a hreflang="en" href="https://bugs.chromium.org/p/chromium/issues/detail?id=629420">bug</a> in Chrome - are expected to have a positive impact. Hint adoption is expected to evolve and the use of `preload` should shift towards its intended purpose - late discovered resources.

Reviewer: ```suggestion
While there are no proposals to add new `rel` directives, improvements from the browser vendors to the current set of resource hints - such as the prioritization <a hreflang="en" href="https://bugs.chromium.org/p/chromium/issues/detail?id=629420">bug</a> in Chrome - are expected to have a positive impact. Hint adoption is expected to evolve, and the use of `preload` should shift towards its intended purpose - late discovered resources.
```

---
#See https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Authors'-Guide#metadata-to-add-at-the-top-of-your-chapters
title: Resource Hints
description: Resource Hints chapter of the 2021 Web Almanac covering adoption of resource hints, their uses, bad practices and their impact on performance.
authors: [kevinfarrugia]
reviewers: [siakaramalegos, tunetheweb, andydavies, samarpanda]
analysts: [nithanaroy]

> editors: [tunetheweb]

Reviewer 2: ```suggestion
editors: [tunetheweb, rviscomi]
```

Reviewer: Happy to remove my name from editor. Can count my edits as part of my reviewer role. I just didn't want this chapter to look like it didn't have an edit if @rviscomi decided it didn't need further edits.

```suggestion
editors: [rviscomi]
```

kevinfarrugia_bio: Kevin Farrugia is a consultant on web performance and software architecture. You can find him blogging on <a hreflang="en" href="https://imkev.dev">imkev.dev</a>.
featured_quote: Resource hints are akin to fine-tuning a race car's engine. They would not turn a slow engine into a fast one, and too many adjustments could break it. Yet, some small tweaks here and there would allow you to maximize it.
featured_stat_1: 44.3%
featured_stat_label_1: Adoption of preload in the top 1,000 websites.
featured_stat_2: 21.5%
featured_stat_label_2: Unused preload hints within the first 3 seconds.
featured_stat_3: 423%
featured_stat_label_3: Growth in adoption of native lazy loading on images.
---

## Introduction

Resource Hints are instructions to the browser that you may use to improve a website's performance. This set of instructions enable you to assist the browser in prioritizing origins or resources which need to be fetched and processed.

Let's take a closer look at how resource hints are implemented, what are the most common pitfalls and what we can do to make sure we are using resource hints as effectively as possible.

> ### The Link directive

Author: I think in this case it should not, but when referring to the `<link>` element it should

> ### The Link directive

Reviewer: I think I agree with @kevinfarrugia 

I did wonder about whether it deserved initial capital letter too, but kind of think of it like a browser API (which would get that treatment).

Plus keeping it as is would keep us consistent with last two years (not essential if it’s wrong but enough to sway the vote if we’re on the borderline).

> ### The Link directive

Reviewer 2: SGTM

```js
const link = document.createElement("link");
link.rel="prefetch";
link.href="https://example.com";

document.head.appendChild(link);
```

Adoption for HTTP headers is significantly lower than having resource hints implemented as part of the document markup; with less than 1.5% of the pages analyzed implementing resource hints through HTTP headers. This is likely attributed to the ease with which they may be added or modified from within the HTML source, when compared to adding an HTTP header on the server.

{{ figure_markup(
  image="http-headers-vs-html-markup.png",
  caption='Popularity of resource hints as HTTP headers and HTML markup.',
  description='Bar chart showing the popularity of resource hints as HTTP headers and HTML markup. HTML markup is used on 91.2% of instances on desktop and 94.1% of instances on mobile. HTTP headers are used on 8.8% of instances on desktop and 5.9% of instances on mobile.',
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTgxCYoH2vDbhTky1qQKEeV034kReHF8JYYq0aDyoo6LG22XL6Tar3dNPsqU1_zyvR6vuFEdMGoF1oP/pubchart?oid=1056457221&format=interactive",
  sheets_gid="1707816066",
  sql_file="http_headers_hints_adoption.sql"
  )
}}

> Using our current [methodology](./methodology), it is not possible to reliably measure resource hints that are added following user-interaction, such as those added through <a hreflang="en" href="https://github.com/GoogleChromeLabs/quicklink">QuickLink</a>, though that particular library featured on less than 0.1% of pages analyzed.

Author: Hmmm, I took it from your awesome CWV Technologies report actually. The query behind it is probably a bit excessive for the scope of this chapter. Not sure if you have a query handy which I could use or whether I should link to the report.

> Using our current [methodology](./methodology), it is not possible to reliably measure resource hints that are added following user-interaction, such as those added through <a hreflang="en" href="https://github.com/GoogleChromeLabs/quicklink">QuickLink</a>, though that particular library featured on less than 0.1% of pages analyzed.

Reviewer 2: Oh in that case I'd suggest adding a quick line of attribution
```suggestion
Using our current [methodology](./methodology), it is not possible to reliably measure resource hints that are added following user-interaction, such as those added through <a hreflang="en" href="https://github.com/GoogleChromeLabs/quicklink">QuickLink</a>, though that particular library featured on less than 0.1% of pages analyzed, according to the [Core Web Vitals Technology Report](https://datastudio.google.com/s/uMbv5CQfW4Q).
```

`modulepreload` is a specialized alternative to `preload` and behaves similarly, however its usage is limited to <a hreflang="en" href="https://html.spec.whatwg.org/multipage/webappapis.html#module-script">module scripts</a>.

## Adoption and trends

{{ figure_markup(
  image="resource-hints-adoption.png",
  caption='Adoption of the link rel attribute.',
  description='Bar chart showing the percent of pages using the link rel attribute values. `dns-prefetch` has 36.4% adoption on mobile and 35.7% on desktop. `preload` has 22.1% adoption on mobile and 22.0% on desktop. `preconnect` has 12.7% adoption on mobile and 12.9% on desktop. `prefetch` has 2.1% adoption on mobile and 2.4% on desktop. `prerender` has 0.1% adoption on mobile and 0.1% on desktop. `modulepreload` has 0.1% adoption on mobile and 0.1% on desktop.',
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTgxCYoH2vDbhTky1qQKEeV034kReHF8JYYq0aDyoo6LG22XL6Tar3dNPsqU1_zyvR6vuFEdMGoF1oP/pubchart?oid=1740212588&format=interactive",
  sheets_gid="2077755325",
  sql_file="hints_adoption.sql"
  )
}}

> The most widely used resource hint is `dns-prefetch` (36.4% on mobile); which is unsurprising, considering it was introduced in <a hreflang="en" href="https://caniuse.com/link-rel-dns-prefetch">2009</a>. With the widespread use of HTTPS, in many cases it should be replaced with `preconnect` (12.7% on mobile). Considering that the `preload` hint is comparatively new, first appearing in Chrome in <a hreflang="en" href="https://groups.google.com/a/chromium.org/g/blink-dev/c/_nu6HlbNQfo/m/XzaLNb1bBgAJ?pli=1">2016</a>, it is the second most widely adopted resource hint (22.1% on mobile) and is seeing constant growth year-on-year—a testament to the importance and flexibility of this directive.

Reviewer 2: If `preconnect` is preferable to `dns-prefetch` this sounds like an opportunity to improve the ecosystem by identifying who the most influential entities are for `dns-prefetch` adoption. Too late to get this analysis into the chapter so I spun up this discussion on the forum: https://discuss.httparchive.org/t/who-is-driving-dns-prefetch-adoption/2263

Reviewer 2: Update: some responses to the thread have suggested that maybe there are still good reasons to keep `dns-prefetch` since `preconnect` is more expensive.

Is the "in many cases" of this paragraph doing too much heavy lifting? Stronger language might be needed to clarify when one hint might be more appropriate than the other.

Author: Originally I had written,
> in many cases it _could_ be replaced
And then changed it to _should_.

Personally I think both have their uses and it isn't like-for-like, but I do believe that in many cases (not all) `dns-prefetch` should be replaced with `preconnect`.

When introducing the `preconnect` hint, I say:
> You are able to use `preconnect` in place of `dns-prefetch` as it gives a greater performance boost; but you must use it sparingly as certificates are usually upwards of 3 KB, which would be competing with bandwidth for other resources. You also want to avoid wasting CPU time opening connections which aren't required for critical resource. Keep in mind that if a connection isn't used within a short period of time (e.g., 10 seconds on Chrome), it would automatically be closed by the browser, wasting any `preconnect` effort.

I was hoping it would help clarify when it should be replaced and when it should not.

Let me know what you think. I agree with the comments on the thread and am trying to convey the same thing.

Reviewer: Given your analysis on that thread it seems both Wordpress and Google Fonts use cases would benefit from this. Maybe you could mention those examples? And then add a caveat about its expense (“as discussed in the pre connection section, there is an added risk/expense…”).

Could also link to that thread?

Author: Updated to:

> With the widespread use of HTTPS, in many cases you should replace it with `preconnect` (12.7% on mobile), if you are certain that you will be connecting to that domain.

Didn't want to link to the thread since the content is already in the chapter.

      <td class="numeric">4</td>
      <td class="numeric">0</td>
      <td class="numeric">4</td>
      <td class="numeric">1</td>
    </tr>
    <tr>
      <td class="numeric">10,000</td>
      <td class="numeric">3</td>
      <td class="numeric">1</td>
      <td class="numeric">4</td>
      <td class="numeric">1</td>
      <td class="numeric">3</td>
      <td class="numeric">1</td>
    </tr>
    <tr>
      <td class="numeric">100,000</td>
      <td class="numeric">2</td>
      <td class="numeric">2</td>
      <td class="numeric">3</td>
      <td class="numeric">1</td>
      <td class="numeric">3</td>
      <td class="numeric">1</td>
    </tr>
    <tr>
      <td class="numeric">1,000,000</td>
      <td class="numeric">2</td>
      <td class="numeric">2</td>
      <td class="numeric">2</td>
      <td class="numeric">1</td>
      <td class="numeric">2</td>
      <td class="numeric">1</td>
    </tr>
    <tr>
      <td class="numeric">all</td>
      <td class="numeric">2</td>
      <td class="numeric">2</td>
      <td class="numeric">1</td>
      <td class="numeric">1</td>
      <td class="numeric">2</td>
      <td class="numeric">1</td>
    </tr>
  </tbody>
</table>

> <figcaption>{{ figure_link(caption='Median resource hint adoption by rank.', sheets_gid="528380369", sql_file="resource_hints_distribution_by_rank.sql") }}</figcaption>

Reviewer 2: ```suggestion
<figcaption>{{ figure_link(caption='Median number of resource hints per page by rank.', sheets_gid="528380369", sql_file="resource_hints_distribution_by_rank.sql") }}</figcaption>
```

</figure>

> Of the sites using resource hints, when comparing the median for the top 1,000 sites to the entire corpus, the top-ranking sites have more resource hints per page. The only hint which observes a different pattern is `prerender`—which has a total of 0 occurrences in the top 1,000 sites.

Reviewer 2: ```suggestion
Of the sites using resource hints, when comparing the median for the top 1,000 sites to the entire corpus, the top-ranking sites have more resource hints per page. The only hint which observes a different pattern is `prerender`, which has a total of 0 occurrences in the top 1,000 sites.
```

## Correlation with Core Web Vitals

{{ figure_markup(
  image="correlation-of-good-cwv-and-preload.png",
  caption='Correlation between good CWV score and number of `rel="preload"` hints',
  description='Scatter plot with trend lines showing the number of pages which have a good CWV score and the number of `rel="preload"` hints on that page. 42.8% of desktop pages, and 30.1% of mobile pages with 0 preload hints have a good CWV score. 21.5% of desktop pages and 11.1% of mobiles page, with 20 preload hints have a good CWV score.',

>   chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTgxCYoH2vDbhTky1qQKEeV034kReHF8JYYq0aDyoo6LG22XL6Tar3dNPsqU1_zyvR6vuFEdMGoF1oP/pubchart?oid=122948841&format=interactive",

Reviewer 2: The x-axis label is "Elements per page" which I think is the label used in the sample chart in the data viz gallery. I think we'll need to update the chart and retake the screenshot.

{{ figure_markup(
  image="correlation-of-good-lcp-and-preload.png",
  caption='Correlation between good LCP score and number of `rel="preload"` hints',
  description='Scatter plot with trend lines showing the number of pages which have a good LCP score and the number of `rel="preload"` hints on that page. 62.5% of desktop pages, and 47.5% of mobile page with 0 preload hints have a good LCP score. 38.2% of desktop pages, and 20.7% of mobile pages with 20 preload hints have a good LCP score.',
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTgxCYoH2vDbhTky1qQKEeV034kReHF8JYYq0aDyoo6LG22XL6Tar3dNPsqU1_zyvR6vuFEdMGoF1oP/pubchart?oid=1380182218&format=interactive",
  sheets_gid="2032682744",
  sql_file="correlation_cwv_preload.sql"
  )
}}

This same observation may be seen on a page's <a hreflang="en" href="https://web.dev/lcp">Largest Contentful Paint</a>, indicating that in many cases, the developer is prioritizing resources which aren't needed to render the LCP element and as a consequence degrading the user experience.

> While this doesn't prove that having preload hints causes a page to get slower, having many does coincide with having slower performance. Every page has its unique requirements and it is impossible to apply a "one size fits all" approach, but in the majority of cases the number of preloaded resources should be kept low and resource prioritization should be delegated to the browser when possible.

Reviewer 2: ```suggestion
While this doesn't prove that having preload hints causes a page to get slower, having many hints does correlate with having slower performance. Every page has its unique requirements and it is impossible to apply a "one size fits all" approach, but in the majority of cases the number of preloaded resources should be kept low and resource prioritization should be delegated to the browser when possible.
```

### The `as` attribute

The `as` attribute should be specified when using `rel="preload"` (or `rel="prefetch"`) to specify the type of resource being downloaded. Applying the correct `as` attribute allows the browser to prioritize the resource more accurately. For example, `preload as="script"` will get a low or medium priority, while `preload as="style"` would be assigned an internal request priority of _Highest_. The `as` attribute is required for caching the resource for future requests and applying the correct <a hreflang="en" href="https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP">Content Security Policy</a>.

{{ figure_markup(
  image="preload-as-attribute-values.png",
  caption='`rel="preload" as` attribute values.',
  description='Bar chart showing the usage of values for rel="preload". script is used 54.6% of the time on mobile and 54.5% on desktop. font is used 21.9% of the time on mobile and 23.5% on desktop. style is used 10.9% of the time on mobile and 11.3% on desktop. fetch is used 9.4% of the time on mobile and 7.5% on desktop. image is used 2.8% of the time on mobile and 2.8% on desktop.',

>   chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTgxCYoH2vDbhTky1qQKEeV034kReHF8JYYq0aDyoo6LG22XL6Tar3dNPsqU1_zyvR6vuFEdMGoF1oP/pubchart?oid=1844544440&format=interactive",

Reviewer 2: Remove "2021" from "Resource Hints 2021" in the subtitle and update the screenshot

#### `fetch`

Preload may be used to initiate a request to retrieve data which you know is critical to the rendering of the page, such as a JSON response or stream.

#### `image`

Preloading images may help improve the LCP score when the image is not included in the initial HTML, such as a CSS `background-image`.

### The `crossorigin` attribute

> The `crossorigin` attribute is used to indicate whether [Cross-Origin Resource Sharing (CORS)](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) must be used when fetching the requested resource. This could apply to any resource type, but it is most commonly associated with font files as they should always be requested using CORS.

Reviewer 2: ```suggestion
The `crossorigin` attribute is used to indicate whether [Cross-Origin Resource Sharing](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) (CORS) must be used when fetching the requested resource. This could apply to any resource type, but it is most commonly associated with font files as they should always be requested using CORS.
```

<figure>
  <table>
    <thead>
      <tr>
        <th>value</th>
        <th>desktop</th>
        <th>mobile</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>not set</td>
        <td class="numeric">66.6%</td>
        <td class="numeric">65.9%</td>
      </tr>
      <tr>
        <td><code>crossorigin</code> (or equivalent)</td>
        <td class="numeric">14.5%</td>
        <td class="numeric">13.5%</td>
      </tr>
      <tr>
        <td><code>use-credentials</code></td>
        <td class="numeric">&lt; 0.1%</td>
        <td class="numeric">&lt; 0.1%</td>
      </tr>
    </tbody>
  </table>

<figcaption>{{ figure_link(caption='<code>rel="preload" crossorigin</code> attribute values.', sheets_gid="1246058294", sql_file="hint_attribute_usage.sql") }}</figcaption>
</figure>

#### `anonymous`

> The default value when no value is specified, `anonymous` will set the credentials flag to <a hreflang="en" href="https://developer.mozilla.org/en-US/docs/Web/Security/Same-origin_policy">`same-origin`</a> and is required when downloading resources protected by CORS. It is also a <a hreflang="en" href="https://drafts.csswg.org/css-fonts/#font-fetching-requirements">requirement</a> when downloading font files—even if they are on the same origin! If you omit the `crossorigin` attribute when the eventual request for the preloaded resource uses CORS, you will end up with a duplicate request since it won't match in the preload cache.

Reviewer 2: ```suggestion
The default value when no value is specified, `anonymous`, will set the credentials flag to <a hreflang="en" href="https://developer.mozilla.org/en-US/docs/Web/Security/Same-origin_policy">`same-origin`</a> and is required when downloading resources protected by CORS. It is also a <a hreflang="en" href="https://drafts.csswg.org/css-fonts/#font-fetching-requirements">requirement</a> when downloading font files—even if they are on the same origin! If you omit the `crossorigin` attribute when the eventual request for the preloaded resource uses CORS, you will end up with a duplicate request since it won't match in the preload cache.
```

Author: I think there shouldn't be a comma here.

Reviewer: I think there should be.

But think it would be better worded as:

```suggestion
The default value when no value is specified is `anonymous` and this value will set the credentials flag to <a hreflang="en" href="https://developer.mozilla.org/en-US/docs/Web/Security/Same-origin_policy">`same-origin`</a>. It is required when downloading resources protected by CORS. It is also a <a hreflang="en" href="https://drafts.csswg.org/css-fonts/#font-fetching-requirements">requirement</a> when downloading font files—even if they are on the same origin! If you omit the `crossorigin` attribute when the eventual request for the preloaded resource uses CORS, you will end up with a duplicate request since it won't match in the preload cache.
```

#### `use-credentials`

> When requesting cross-origin resources which require authentication, for example through the use of cookies, client certificates or the `Authorization` header; setting the `crossorigin="use-credentials"` attribute will include this data in the request and allowing the server to respond to the request so that the resource may be preloaded. This is not a common scenario with 0.1% usage, however if your page content is dependent on an authenticated status, it could be used to initiate an early fetch request to get the login status.

Reviewer 2: ```suggestion
When requesting cross-origin resources which require authentication, for example through the use of cookies, client certificates or the `Authorization` header; setting the `crossorigin="use-credentials"` attribute will include this data in the request and allow the server to respond to the request so that the resource may be preloaded. This is not a common scenario with 0.1% usage, however if your page content is dependent on an authenticated status, it could be used to initiate an early fetch request to get the login status.
```

### The `media` attribute

> An oft-neglected feature available to `rel="preload"`, is the ability to specify media queries through the `media` attribute—with less than 4% of all preloads using this attribute. The `media` attribute accepts media queries allowing you to target the media type and specific browser features, such as viewport width. As an example, the `media` attribute would allow you to preload a low-resolution image on devices with a narrow viewport and a full-sized image on devices with a large viewport.

Reviewer 2: ```suggestion
An oft-neglected feature available to `rel="preload"` is the ability to specify media queries through the `media` attribute—with less than 4% of all preloads using this attribute. The `media` attribute accepts media queries allowing you to target the media type and specific browser features, such as viewport width. As an example, the `media` attribute would allow you to preload a low-resolution image on devices with a narrow viewport and a full-sized image on devices with a large viewport.
```

In addition to the `media` attribute, the `<link>` element supports `imagesrcset` and `imagesizes` attributes which correspond to the `srcset` and `sizes` attributes on `<img>` elements. Using these attributes, you can use the same resource selection criteria that you would use on your image. Unfortunately, their adoption is very low (less than 1%); most likely owing to the lack of <a hreflang="en" href="https://caniuse.com/mdn-html_elements_link_imagesizes">support</a> on Safari.

<p class="note">Note: The `media` attribute is not available on all `<link>` elements as the spec suggests, but it is only available on `rel="preload"`.</p>

### Bad practices

Owing to the versatility of `rel="preload"`, there isn't a clear set of rules dictating how to implement the preload hint, but we can learn a lot from our mistakes and understand how to avoid them.

#### Unused preload

We have already seen that there is a negative correlation between a website's performance and the number of preload hints. This relationship may be influenced by two factors:
- Incorrect preloads
- Unused preloads

##### Incorrect preloads

> When you preload a resource which is not as important as the other resources which the browser would have otherwise prioritized. We are unable to measure the extent of incorrect preloads as you would need to A/B test the page with and without each hint.

Reviewer 2: Consider rephrasing. The first sentence is incomplete.

> ##### Unused preloads

Reviewer 2: It feels strange that "Unused preloads" is a subsection within "Unused preload". Could we remove these subheadings and nest the content directly under "Unused preload"? May also need to rephrase each section to clarify which bad practice it's referring to.

Author: Rephrased section.

> When you preload a resource which is not needed within the first few seconds of the page loading.

Reviewer 2: Complete sentence

{{ figure_markup(
  caption='Percent of unused preload hints within the first 3 seconds.',
  content="21.5%",
  classes="big-number",
  sheets_gid="2013605735",
  sql_file="consoleLog_unused_preload.sql"
)
}}

In such cases, the preload hint is regressing the website's performance, as you are instructing the browser to download and prioritize files or resources which are not needed immediately—or even not needed at all. This is one of the challenges when using resource hints, as they require regular maintenance and automating the process opens the door to allow such issues to creep in.

> {{ figure_markup(

Reviewer 2: This figure doesn't seem to be discussed. Add a paragraph below with your takeaway.

  sheets_gid="699889350",
  sql_file="consoleLog_incorrect_crossorigin_type.sql"
  )
}}

More than half (63.6%) of the cases when the `crossorigin` attribute on the `rel="preload"` hint is either missing or incorrect, are linked to the preloading of font files, with a total of 14,818 instances across the dataset.

#### Invalid `as` attribute

The `as` attribute plays an important role when preloading your resources and getting this wrong may result in downloading the same resource twice. On most browsers, specifying an unrecognized `as` attribute will ignore the preload. The supported values are `audio`, `document`, `embed`, `fetch`, `font`, `image`, `object`, `script`, `style`, `track`, `worker` and `video`.

There are 17,861 cases of unrecognized values, with the most frequent error being omitting it completely; while the most common invalid as values are `other` and `stylesheet` (the correct value is `style`).

{{ figure_markup(
  caption='Pages incorrectly used `as="stylesheet"` instead of `"style"`',
  content="1,114",
  classes="big-number",
  sheets_gid="1681733418",
  sql_file="consoleLog_unused_preload.sql"
)
}}

> When using an incorrect `as` attribute value—as opposed to unrecognized value, such as using `style` instead of `script`; the browser will duplicate the file download as the request won't match the resource stored in the preload cache.

Reviewer 2: ```suggestion
When using an incorrect `as` attribute value—as opposed to unrecognized value, such as using `style` instead of `script`—the browser will duplicate the file download as the request won't match the resource stored in the preload cache.
```

<p class="note">Note: While <code>video</code> is included in the spec, it isn't supported by any browser and would be treated as an invalid value and ignored.</p>

#### Unused font files

> More than 5% of pages which preload font files, preload more font files than needed. When preloading font files, all browsers which support `preload` also support `.woff2`. This means that, assuming that the `.woff2` font files are available, it is not necessary to preload older formats, including `.woff`.

Reviewer 2: ```suggestion
More than 5% of pages which preload font files preload more font files than needed. When preloading font files, all browsers which support `preload` also support `.woff2`. This means that, assuming that the `.woff2` font files are available, it is not necessary to preload older formats, including `.woff`.
```

Author: Are you sure this comma should be removed?

Reviewer: I think it should be removed. They are not separate clauses but one is a direct continuation of the other.

## Third parties

> You can use resource hints to connect to, or download resources from, both first and [third parties](./third-parties). While `dns-prefetch` and `preconnect` are only useful when connecting to different origins, including subdomains; `preload` and `prefetch` may be used for both resources on the same origin and resources hosted by third parties.

Reviewer 2: I'd recommend rephrasing this paragraph to more explicitly call out the Third Parties chapter. As written, it's not clear that clicking the link will take me to another chapter.

For example, at the end of the paragraph (or at the end of this section, as it's all about 3P) you could say something like "To learn more about the state of third parties, check out the [Third Parties](./third-parties) chapter."

> You can use resource hints to connect to, or download resources from, both first and [third parties](./third-parties). While `dns-prefetch` and `preconnect` are only useful when connecting to different origins, including subdomains; `preload` and `prefetch` may be used for both resources on the same origin and resources hosted by third parties.

Reviewer 2: ```suggestion
You can use resource hints to connect to, or download resources from, both first and [third parties](./third-parties). While `dns-prefetch` and `preconnect` are only useful when connecting to different origins, including subdomains, `preload` and `prefetch` may be used for both resources on the same origin and resources hosted by third parties.
```

When considering which resource hints you should use for third-party resources, you need to evaluate the priority and role of each third party on your application's loading experience and whether the costs are justified.

> Prioritizing third-party resources over your own content is potentially a warning sign, however there are cases when this is recommended. As an example, if we look at cookie notice scripts; which are required in the European Union by <a href="https://en.wikipedia.org/wiki/General_Data_Protection_Regulation">General Data Protection Regulation</a>; these are usually accompanied by a `dns-prefetch` or `preconnect` script as they are highly obtrusive to the user experience and also a prerequisite for some site functions, such as serving personalized ads.

Reviewer 2: ```suggestion
Prioritizing third-party resources over your own content is potentially a warning sign, however there are cases when this is recommended. As an example, if we look at cookie notice scripts—which are required in the European Union by <a href="https://en.wikipedia.org/wiki/General_Data_Protection_Regulation">General Data Protection Regulation</a>—these are usually accompanied by a `dns-prefetch` or `preconnect` hint as they are highly obtrusive to the user experience and also a prerequisite for some site functions, such as serving personalized ads.
```

        <td class="numeric">12.2%</td>
        <td class="numeric">14.0%</td>
      </tr>
      <tr>
        <td>`www.google-analytics.com`</td>
        <td class="numeric">1.3%</td>
        <td class="numeric">4.1%</td>
        <td class="numeric">0.2%</td>
        <td class="numeric">5.5%</td>
      </tr>
      <tr>
        <td>`www.googletagmanager.com`</td>
        <td class="numeric">1.9%</td>
        <td class="numeric">3.1%</td>
        <td class="numeric">0.3%</td>
        <td class="numeric">5.3%</td>
      </tr>
      <tr>
        <td>`siteassets.parastorage.com`</td>
        <td class="numeric">-</td>
        <td class="numeric">-</td>
        <td class="numeric">5.2%</td>
        <td class="numeric">5.2%</td>
      </tr>
      <tr>
        <td>`images.squarespace-cdn.com`</td>
        <td class="numeric">-</td>
        <td class="numeric">4.5%</td>
        <td class="numeric">-</td>
        <td class="numeric">4.5%</td>
      </tr>
      <tr>
        <td>`ajax.googleapis.com`</td>
        <td class="numeric">2.3%</td>
        <td class="numeric">1.8%</td>
        <td class="numeric">0.4%</td>
        <td class="numeric">4.5%</td>
      </tr>
    </tbody>
  </table>

> <figcaption>{{ figure_link(caption="Most popular third parties connected to using resource hints.", sheets_gid="1254656281", sql_file="preload_host_by_url.sql") }}</figcaption>

Reviewer 2: ```suggestion
<figcaption>{{ figure_link(caption="Most popular third-party connections using resource hints.", sheets_gid="1254656281", sql_file="preload_host_by_url.sql") }}</figcaption>
```

</figure>

Analyzing the table above, 42.7% of all pages which include a `preload` hint are preloading resources hosted on adservice.google.com, while s.w.org (WordPress) is the most popular domain for `dns-prefetch` and `fonts.gstatic.com` for the `preconnect` directive.

{{ figure_markup(
  image="google-fonts.png",
  alt="Google Fonts instructions to embed a font.",

>   caption='Google Fonts instructions to preconnect to fonts.gstatic.com and fonts.googleapis.com. From <a hreflang="en" href="https://fonts.google.com/">Google Fonts</a>.',

Reviewer 2: ```suggestion
  caption='Google Fonts instructions to preconnect to fonts.gstatic.com and fonts.googleapis.com. (Source: <a hreflang="en" href="https://fonts.google.com/">Google Fonts</a>)',
```

  description='Code snippet titled "Use on the web" and subtitled with "To embed a font, copy the code into the `<head>` of your html". Code snippet states `<link rel="preconnect" href="https://fonts.googleapis.com"><link rel="preconnect" href="https://fonts.gstatic.com" crossorigin><link href="https://fonts.googleapis.com/css2?family=Roboto&display=swap" rel="stylesheet">`',
  width=536,

>   height=510

Reviewer 2: This figure appears very zoomed in. Should the dimensions be halved?

Author: Should I resize the original image or change the width/height attributes to preserve the higher-resolution image?

Reviewer: just change height and width and leave image itself alone. In same way we say charts are 600px when they are actually 1200px.

  )
}}

Google Fonts now includes instructions to `preconnect` to both the fonts.gstatic.com origin and fonts.googleapis.com, which is usually good practice to offset the impact of these late discovered resources.

## Native lazy-loading

Lazy-loading refers to the technique to defer downloading a resource—in this case an image or iframe—until it is needed or visible within the viewport. Native lazy-loading refers to the ability to specify this in the HTML with a `loading="lazy"` attribute, rather than having to use a JavaScript library to handle this. Native image and iframe lazy-loading have been standardized in 2019 and since then their adoption—especially for images—has grown exponentially.

`loading="lazy"` for images is supported on most major browsers. On Safari, it is marked as <a hreflang="en" href="https://bugs.webkit.org/show_bug.cgi?id=200764">in progress</a> and is available behind a flag, but not yet enabled by default.

Lazy-loading of iframes is supported on Chrome, once again behind a flag on Safari but <a hreflang="en" href="https://bugzilla.mozilla.org/show_bug.cgi?id=1622090">not yet supported on Firefox</a>.

> Browsers which do not support the `loading` attribute, will simply ignore it—making it safe to add without unwanted side-effects. JavaScript based alternatives, such as <a hreflang="en" href="https://github.com/aFarkas/lazysizes">lazysizes</a> may still be used, however considering that full browser support is around the corner, it may not be worth adding to a project at this stage.

Reviewer 2: ```suggestion
Browsers which do not support the `loading` attribute will simply ignore it—making it safe to add without unwanted side-effects. JavaScript based alternatives, such as <a hreflang="en" href="https://github.com/aFarkas/lazysizes">lazysizes</a> may still be used, however considering that full browser support is around the corner, it may not be worth adding to a project at this stage.
```

{{ figure_markup(
  image="adoption-of-loading-lazy-on-img.png",
  caption='The percent of pages that have the `loading="lazy"` attribute on `img` elements.',
  description='Time series showing the percent of pages that have the `loading="lazy"` attribute on `img` elements on desktop and mobile. At the start of 2019 the adoption rate is 0% on desktop and 0% on mobile. At the start of 2020 the adoption rate is 0% on desktop and 0% on mobile. At the start of 2021 the adoption rate is 14.2% on desktop and 13.5% on mobile. As of July 2021 the adoption rate is 17.9% on desktop and 17.8% on mobile.',
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTgxCYoH2vDbhTky1qQKEeV034kReHF8JYYq0aDyoo6LG22XL6Tar3dNPsqU1_zyvR6vuFEdMGoF1oP/pubchart?oid=1314627953&format=interactive",
  sheets_gid="157636784",
  sql_file="imgLazy.sql"
  )
}}

> The number of pages using `loading="lazy"` has grown from 4.2% in 2020 to 17.8% by the time of our analysis. That's a whopping 423% growth! This rapid growth is extraordinary and is likely driven by two key elements. The ease with which it could be added to pages without cross-browser compatibility issues, and the frameworks or technologies powering these websites. In WordPress 5.5, <a hreflang="en" href="https://make.wordpress.org/core/2020/07/14/lazy-loading-images-in-5-5/">lazy-loading images became the default implementation</a>, supercharging the adoption rate of `loading="lazy"`, with WordPress sites now making up <a hreflang="en" href="https://web.dev/lcp-lazy-loading/">84%</a> of all pages which use native image lazy-loading.

Reviewer 2: ```suggestion
The percent of pages using `loading="lazy"` has grown from 4.2% in 2020 to 17.8% by the time of our analysis. That's a whopping 423% growth! This rapid growth is extraordinary and is likely driven by two key elements: the ease with which it could be added to pages without cross-browser compatibility issues, and the frameworks or technologies powering these websites. In WordPress 5.5, <a hreflang="en" href="https://make.wordpress.org/core/2020/07/14/lazy-loading-images-in-5-5/">lazy-loading images became the default implementation</a>, supercharging the adoption rate of `loading="lazy"`, with WordPress sites now making up <a hreflang="en" href="https://web.dev/lcp-lazy-loading/">84%</a> of all pages which use native image lazy-loading.
```

> A <a hreflang="en" href="https://web.dev/lcp-lazy-loading/">study</a> on the load times for pages which use lazy-loading, indicated that pages which use lazy-loading tend to have a worse LCP performance, possibly caused by overusing the lazy-loading attribute. This is increasingly significant on the LCP element, which shouldn't be lazy-loaded. If you are using `loading="lazy"`, you should check that the lazily-loaded images are below the fold and more critically, that the [LCP element is not lazy-loaded](./performance). More than 60% of all lazy-load images are actually within the initial viewport and shouldn't be lazy-loaded.)

Reviewer 2: ```suggestion
A <a hreflang="en" href="https://web.dev/lcp-lazy-loading/">study</a> on the load times for pages which use lazy-loading indicated that pages which use lazy-loading tend to have a worse LCP performance, possibly caused by overusing the lazy-loading attribute. This is increasingly significant on the LCP element, which shouldn't be lazy-loaded. If you are using `loading="lazy"`, you should check that the lazily-loaded images are below the fold and more critically, that the [LCP element is not lazy-loaded](./performance). More than 60% of all lazy-load images are actually within the initial viewport and shouldn't be lazy-loaded.
```

> A <a hreflang="en" href="https://web.dev/lcp-lazy-loading/">study</a> on the load times for pages which use lazy-loading, indicated that pages which use lazy-loading tend to have a worse LCP performance, possibly caused by overusing the lazy-loading attribute. This is increasingly significant on the LCP element, which shouldn't be lazy-loaded. If you are using `loading="lazy"`, you should check that the lazily-loaded images are below the fold and more critically, that the [LCP element is not lazy-loaded](./performance). More than 60% of all lazy-load images are actually within the initial viewport and shouldn't be lazy-loaded.)

Reviewer 2: Similar to my Third Parties chapter feedback, more explicitly link to the Performance chapter.

> A <a hreflang="en" href="https://web.dev/lcp-lazy-loading/">study</a> on the load times for pages which use lazy-loading, indicated that pages which use lazy-loading tend to have a worse LCP performance, possibly caused by overusing the lazy-loading attribute. This is increasingly significant on the LCP element, which shouldn't be lazy-loaded. If you are using `loading="lazy"`, you should check that the lazily-loaded images are below the fold and more critically, that the [LCP element is not lazy-loaded](./performance). More than 60% of all lazy-load images are actually within the initial viewport and shouldn't be lazy-loaded.)

Reviewer 2: Also the 60% stat comes before figure 18 so at first I was confused that it was from the Performance chapter linked immediately prior. Consider moving the figure up before this paragraph or moving the 60% reference to below the figure.

{{ figure_markup(
  image="lazy-loaded-images.png",
  caption='Percent of `img` elements with `loading="lazy"` which are in the initial viewport.',
  description='Bar chart showing the percent of `img` elements with `loading="lazy"` which are within the initial viewport. 61.5% on mobile and 63.1% on desktop of lazy-loaded images are within the initial viewport. 38.5% on mobile and 36.9% on desktop of lazy-loaded images are not within the initial viewport.',
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTgxCYoH2vDbhTky1qQKEeV034kReHF8JYYq0aDyoo6LG22XL6Tar3dNPsqU1_zyvR6vuFEdMGoF1oP/pubchart?oid=977858704&format=interactive",
  sheets_gid="1468369891",
  sql_file="lazy_viewport_images.sql"
  )
}}

> The likelihood of a page containing at least one iframe is much lower than for that containing an image with only 2.6% of pages containing an iframe taking advantage of native lazy-loading. The benefits of lazy-loading an iframe are potentially important, as an iframe could initiate further requests to download even more resources, including scripts and images. This is especially true when using embeds, such as YouTube or Twitter embeds. Similarly, to when deciding the loading strategy for an image, you must check whether the iframe is shown within the initial viewport or not. If it isn't, then it is usually safe to add `loading="lazy"` to the `<iframe>` element to benefit from a reduced initial load and boost performance.

Reviewer 2: I'd suggest putting the 2.6% stat in a figure so its source material is more easily discoverable. Given that this paragraph is right below figure 18, I was initially confused at how the two are related. A dedicated figure above this paragraph would help address that.

> The likelihood of a page containing at least one iframe is much lower than for that containing an image with only 2.6% of pages containing an iframe taking advantage of native lazy-loading. The benefits of lazy-loading an iframe are potentially important, as an iframe could initiate further requests to download even more resources, including scripts and images. This is especially true when using embeds, such as YouTube or Twitter embeds. Similarly, to when deciding the loading strategy for an image, you must check whether the iframe is shown within the initial viewport or not. If it isn't, then it is usually safe to add `loading="lazy"` to the `<iframe>` element to benefit from a reduced initial load and boost performance.

Reviewer 2: ```suggestion
The likelihood of a page containing at least one iframe is much lower than for that containing an image with only 2.6% of pages containing an iframe taking advantage of native lazy-loading. The benefits of lazy-loading an iframe are potentially important, as an iframe could initiate further requests to download even more resources, including scripts and images. This is especially true when using embeds, such as YouTube or Twitter embeds. Similarly, when deciding the loading strategy for an image, you must check whether the iframe is shown within the initial viewport or not. If it isn't, then it is usually safe to add `loading="lazy"` to the `<iframe>` element to benefit from a reduced initial load and boost performance.
```

> ## HTTP/2 Server Push

Reviewer 2: Link to the HTTP chapter somewhere in this section?

Unfortunately, HTTP/2 push has been disappointing, with little evidence that it provides the performance boost promised compared to the risk of over pushing resources that either the browser already has, or that are of less importance than resources the browser requests.

So, while the technology is widely available, overcoming these obstacles makes it highly unpopular—with less than 1% adoption. Chrome has also filed an <a hreflang="en" href="https://lists.w3.org/Archives/Public/ietf-http-wg/2019JulSep/0078.html">Intent to Remove</a> that is paused until a testable implementation of 103 Early Hints (covered next) is available. Chrome <a hreflang="en" href="https://github.com/httpwg/http2-spec/issues/786#issuecomment-724371629">does not support</a> Server Push on HTTP/3 either.

## Future

> While there are no proposals to add new `rel` directives, improvements from the browser vendors to the current set of resource hints—such as the prioritization <a hreflang="en" href="https://bugs.chromium.org/p/chromium/issues/detail?id=629420">bug</a> in Chrome—are expected to have a positive impact. Hint adoption is expected to evolve, and the use of `preload` should shift towards its intended purpose—late discovered resources.

Reviewer 2: ```suggestion
While there are no proposals to add new `rel` directives, improvements from the browser vendors to the current set of resource hints—such as the prioritization <a hreflang="en" href="https://bugs.chromium.org/p/chromium/issues/detail?id=629420">bug</a> in Chrome—are expected to have a positive impact. Hint adoption is expected to evolve, and the use of `preload` should shift towards its intended purpose: late discovered resources.
```

> Additionally, two proposals, 103 Early Hints and Priority Hints are expected to be made available soon, with experimental support already available on Chrome.

Reviewer 2: ```suggestion
Additionally, two proposals, 103 Early Hints and Priority Hints, are expected to be made available soon, with experimental support already available on Chrome.
```

---
#See https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Authors'-Guide#metadata-to-add-at-the-top-of-your-chapters
title: Resource Hints
description: Resource Hints chapter of the 2021 Web Almanac covering adoption of resource hints, their uses, bad practices, and their impact on performance.
authors: [kevinfarrugia]

> reviewers: [siakaramalegos, tunetheweb, andydavies, samarpanda]

Reviewer 2: Adding @westonruter as a reviewer.

Could you add his contributor info to `2021.json` as well?

```json

    "westonruter": {
      "name": "Weston Ruter",
      "teams": [
        "reviewers"
      ],
      "avatar_url": "https://avatars.githubusercontent.com/u/134745?v=4&s=200",
      "website": "https://weston.ruter.net/",
      "github": "westonruter",
      "twitter": "westonruter"
    },
```

```suggestion
reviewers: [siakaramalegos, tunetheweb, andydavies, samarpanda, westonruter]
```

Author: Added

    <tbody>
      <tr>
        <td>not set</td>
        <td class="numeric">66.6%</td>
        <td class="numeric">65.9%</td>
      </tr>
      <tr>
        <td><code>crossorigin</code> (or equivalent)</td>
        <td class="numeric">14.5%</td>
        <td class="numeric">13.5%</td>
      </tr>
      <tr>
        <td><code>use-credentials</code></td>

>         <td class="numeric">0.1%</td>

Author: @tunetheweb is this intentionally rounding up?

Reviewer: Good spot. What do you think about `< 0.1%` to show it is used and not 0?

  )
}}

#### `script`

`script` is the most common value by a significant margin. `<script>` elements are usually discovered early as they are embedded in the initial HTML document, but it is a common practice to place `<script>` elements before the closing `<body>` tag. Since HTML is parsed sequentially, this means that the scripts will be discovered after the DOM is downloaded and parsed—and with more websites dependent on JavaScript frameworks, the necessity to have JavaScript load early has increased. The downside is that JavaScript resources would be prioritized over the other resources discovered within the HTML document, including images and stylesheets, possibly compromising the user experience.

#### `font`

The second most commonly preloaded resource is the `font`, which is a late-discovered resource since the browser will only download a font file after the layout phase when the browser knows that the font will be rendered on the page.

#### `style`

> Stylesheets are ordinarily embedded in the document's `<head>` and discovered early during the document parsing. Additionally, as stylesheets are render-blocking resources they are assigned the _highest_ request priority. This should make preloading stylesheets unnecessary, but it is sometimes required to re-prioritize the requests. A <a hreflang="en" href="https://bugs.chromium.org/p/chromium/issues/detail?id=629420">bug</a> in Google Chrome (fixed in Chrome 95) prioritizes preloaded resources ahead of other higher-priority resources discovered by the preload scanner, including CSS files. Preloading the stylesheet will restore its _highest_ priority. Another instance when stylesheets are preloaded is when they are not downloaded directly from the HTML document, such as the <a hreflang="en" href="https://www.filamentgroup.com/lab/async-css.html">asynchronous CSS</a> "hack" which uses an `onload` event to avoid render-blocking the page with non-critical CSS.

Author: Here I used "Highest" (propercase) because I'm referring to the enumeration. Not sure which is more correct.

Reviewer: Yeah I struggled over this one myself. I think it reads weird since we haven't explained that's a specific term, and works fine with lowercase too. So changed it.

Up for changing it back but think we should make it clear. "they are assigned an internal request priority of _Highest_" or something like that. Then can leave second one with uppercase without need of a further explanation.

WDYT?

Author: Updated to introduce it earlier on and change both of these back to uppercase.

  )
}}

> More than half of the cases when the `crossorigin` attribute on the `rel="preload"` hint is either missing or incorrect, are linked to the preloading of font files, with a total of 14,818 instances across the dataset.

Author: If you prefer being more specific (instead of "more than half"), it is 63.6% on mobile.

Reviewer: I think that's good! Use it. More than half to me suggests %52 - i.e. it's just more than half, or you would have said so.

Reviewer: "More than half (63.6%) of the cases..." works for me - best of both worlds!

<figure>
  <table>

>   <thead>

Author: Previous table was for desktop. Updated to show mobile.

      <td class="numeric">2.7%</td>
      <td class="numeric">1.0%</td>
      <td class="numeric">0.1%</td>
      <td class="numeric">3.8%</td>
    </tr>
    <tr>
      <td>images.squarespace-cdn.com</td>
      <td class="numeric">-</td>
      <td class="numeric">3.5%</td>
      <td class="numeric">-</td>
      <td class="numeric">3.5%</td>
    </tr>
    <tr>
      <td>cdnjs.cloudflare.com</td>
      <td class="numeric">1.6%</td>
      <td class="numeric">1.0%</td>
      <td class="numeric">0.4%</td>
      <td class="numeric">2.9%</td>
    </tr>
    <tr>
      <td>monorail-edge.shopifysvc.com</td>
      <td class="numeric">2.0%</td>
      <td class="numeric">0.8%</td>
      <td class="numeric">-</td>
      <td class="numeric">2.8%</td>
    </tr>
    <tr>
      <td>fonts.shopifycdn.com</td>
      <td class="numeric">-</td>
      <td class="numeric">1.1%</td>
      <td class="numeric">1.0%</td>
      <td class="numeric">2.1%</td>
    </tr>
  </tbody>
</table>

<figcaption>{{ figure_link(caption="Most popular third-party connections using resource hints on mobile devices.", sheets_gid="1254656281", sql_file="preload_host_by_url.sql") }}</figcaption>
</figure>

> Analyzing the table above, 36.7% of all pages which include a `preload` hint are preloading resources hosted on adservice.google.com. s.w.org (WordPress) is the most popular domain for `dns-prefetch` and is added by default on new WordPress.org installations. Google Fonts related services on `fonts.gstatic.com` and `fonts.googleapis.com` are the two most popular hosts for the `preconnect` directive.

Author: Trying to get confirmation on this statement.

> is added by default on new WordPress.org installations

Author: https://core.trac.wordpress.org/ticket/34292

> By default, wp_resource_hints() prints hints for "s.w.org" (the WordPress.org CDN) and for all scripts and styles which are enqueued from external hosts.

Reviewer 3: The `s.w.org` domain is the CDN for serving emoji SVG images (provided by Twemoji) when WordPress detects (via its inlined `emoji-loader.js`) that the browser doesn't support native emoji characters. This CDN has been used since WordPress 4.6. 

```suggestion
Analyzing the table above, 36.7% of all pages which include a `preload` hint are preloading resources hosted on adservice.google.com. The s.w.org host is the most popular domain for `dns-prefetch` and is used on WordPress sites (since version 4.6) for the loading of SVG images from its Twemoji CDN, when the browser is detected to not support native emoji characters. Google Fonts related services on `fonts.gstatic.com` and `fonts.googleapis.com` are the two most popular hosts for the `preconnect` directive.
```

## HTTP/2 Server Push

HTTP/2 supports a technology called _Server Push_ that preemptively pushes a resource it expects the client will be requesting. As the server is pushing the resource instead of informing the client that it should request it, cache-management becomes complex and, in some cases, the pushed resources would even delay the delivery of the HTML, which is critical for discovering all resources required to load the page.

Unfortunately, HTTP/2 push has been disappointing, with little evidence that it provides the performance boost promised compared to the risk of over pushing resources that either the browser already has, or that are of less importance than resources the browser requests.

> So, while the technology is widely available, overcoming these obstacles makes it highly unpopular—with less than 1% adoption. Chrome has also filed an <a hreflang="en" href="https://lists.w3.org/Archives/Public/ietf-http-wg/2019JulSep/0078.html">Intent to Remove</a> that is paused until a testable implementation of 103 Early Hints (covered next) is available. Chrome <a hreflang="en" href="https://github.com/httpwg/http2-spec/issues/786#issuecomment-724371629">does not support</a> Server Push on HTTP/3 either.

Author: Should we do an internal link for "covered next"?

Reviewer: I don't think so as it literally is next. If it was a good bit away for this, then Id' agree with you.

So, while the technology is widely available, overcoming these obstacles makes it highly unpopular—with less than 1% adoption. Chrome has also filed an <a hreflang="en" href="https://lists.w3.org/Archives/Public/ietf-http-wg/2019JulSep/0078.html">Intent to Remove</a> that is paused until a testable implementation of 103 Early Hints (covered next) is available. Chrome <a hreflang="en" href="https://github.com/httpwg/http2-spec/issues/786#issuecomment-724371629">does not support</a> Server Push on HTTP/3 either.

> You can read more about HTTP, HTTP/2, and HTTP/3 in the [HTTP Chapter](./http).

Reviewer 2: ```suggestion
You can read more about HTTP, HTTP/2, and HTTP/3 in the [HTTP](./http) chapter.
```
